### ISSUE TITLE
Grape ~> 1.3.0 does not set Content-Length header anymore

### ISSUE THREAD

**Issue by nbulaj** [ruby-grape/grape#2051]
**EDITED**
`grape < 1.3.0` sets `Content-Length` header
`grape ~> 1.3.0` does not set it anymore 

the regression is related to the change in https://github.com/rack/rack/commit/8c62821f4a464858a6b6ca3c3966ec308d2bb53e

**ORIGINAL POST**
There are some changes inroduced in 1.3.0 (and still valid for 1.3.2) that broke our production app after update.

We have _app1_ (Rails + Grape), _app2_ (Sinatra), _frontend app_. Frontend app communicates with _app2_ which in turn proxy requests to app1 using `curb` gem and just return responses. No specific magic here. When we used Grape 1.2.5 - everything was OK. But when we just changed version number in Gemfile, done `bundle update grape` and run the app with 1.3.0 - frontend app started to show infinite loader for requests. 

While debugging I found that Grape API started to respond to `curb` gem request with a header - Transfer-Encoding `chunked` and without content-length so frontend app just waits while request finishes for a super-long time. 

This changes that do the stuff introduced here - https://github.com/ruby-grape/grape/pull/1956

When I use Grape version before PR merged:

```ruby
gem "grape", git: "https://github.com/ruby-grape/grape.git", ref: "9f786adbffa25574916e5d5504dacb4f5e79c6bb"
```

Everything works. Then I change it to (git commit when PR was merged):

```ruby
gem "grape", git: "https://github.com/ruby-grape/grape.git", ref: "547bb71471a9273a4c2f8023332a0d5123d3eafa"
```

And  _app2_ starts to receive response headers with Transfer-Encoding `chunked` and never finish the request. Should I create an issue for that? I don't know if could create a reproducable example..

Reproducable example could be found here: https://gist.github.com/nbulaj/bd89ba366290033bf71c0f9b60927709

Also a possible reason of the regression posted here: https://github.com/ruby-grape/grape/issues/2030#issuecomment-624784743

**Comment by dm1try** [comment]
> There are some changes inroduced in 1.3.0 (and still valid for 1.3.2) that broke our production app after update

@nbulaj , which version of `rack` had been used in the examples above?

TLDR
```ruby
require 'bundler/inline'
gemfile(true) do
  source 'https://rubygems.org'
  gem 'rack', '2.0.8'
end
pp Rack::Response.new(["hello from rack 2.0.8"], 200, {})
```
```
#<Rack::Response:0x00007fa2c8595a48
 ...
 @body=["hello from rack 2.0.8"],
 @header={"Content-Length"=>"21"},
...
```
```ruby
require 'bundler/inline'
gemfile(true) do
  source 'https://rubygems.org'
  gem 'rack', '2.1.0'
end
pp Rack::Response.new(["hello from rack 2.1.0"], 200, {})
```
```
#<Rack::Response:0x00007fda2b0f7438
 ...
 @body=["hello from rack 2.1.0"],
 @header={}
  ...
```

The changes in #1956 were introduced to fix the grape invalid implementation of rack spec. see https://github.com/ruby-grape/grape/issues/1966; it means any version of grape`< 1.3` does not work with rack `=> 2.1`.
So upgrading `grape` to `1.3` forces you to upgrade `rack` to version `~> 2.1`.
Some changes introduced in `rack 2.1.0`(https://github.com/rack/rack/commit/8c62821f4a464858a6b6ca3c3966ec308d2bb53e) lead to the behavior where you do not have "Content-Length" header anymore(I assume `grape` implicitly had been relied on it)

**Comment by dm1try** [comment]
Anyway, your proxy client should support chunked transferring if it implements `HTTP/1.1` otherwise you can come across another errors along the stack. Also you can explicitly set protocol version to 1.0 to have a guard from an application server, see https://github.com/puma/puma/blob/4.3.3/lib/puma/server.rb#L799-L805

**Comment by nbulaj** [comment]
>  which version of rack had been used in the examples above?

This is gemfile.lock content:

```ruby
    rack (2.2.2)
    rack-accept (0.4.5)
      rack (>= 0.4)
    rack-protection (2.0.8.1)
      rack
    rack-test (1.1.0)
      rack (>= 1.0, < 3)
```

> it means any version of grape< 1.3 does not work with rack => 2.1.
> So upgrading grape to 1.3 forces you to upgrade rack to version ~> 2.1

Whooa, I thought for such cases there is `gem.add_runtime_dependency` :thinking: 

> Anyway, your proxy client should support chunked transferring if it implements HTTP/1.1

The code you see in my gist is a production code. It really uses `Curl::Easy` to send the request to the main API and then just returns the response (body, status, headers) to the world (frontend app). The strange thing for me is that with 1.2.5 everything was OK and with the changes pointed above it doesn't work anymore :(

**Comment by dm1try** [comment]
>Whooa, I thought for such cases there is gem.add_runtime_dependency

Nope, for the different(where you have a clear predictions about requirements; re-read about the bug and the changes in #1956).

> The code you see in my gist is a production code..

Again, the code has a bug, it does not implement streaming mechanism that available in HTTP/1.1

>It really uses Curl::Easy to send the request to the main API and then just returns the response (body, status, headers) to the world (frontend app).

no, it does not return the response to the world, it sets the result expected by `sinatra` interface, for a usual not-streamed response. it is important, because from this point if you want forward streamed response you should use different `sinatra` interface.

set `HTTP/1.0` in the proxy client to tell the server that the streaming is not supported(and many other features 😑 ):
```
    curl.version = Curl::HTTP_1_0
    curl.perform
```
and check this out, it should work for any stack, including the latest `grape`.

> The strange thing for me is that with 1.2.5 everything was OK and with the changes pointed above it doesn't work anymore

it happens

**Comment by dm1try** [comment]
I've updated the title and the original post to narrow down the problem.
btw, there is no spec that `grape` should set `Content-Length`, so it should be discussed first)

**Comment by nbulaj** [comment]
> Again, the code has a bug, it does not implement streaming mechanism that available in HTTP/1.1

> it is important, because from this point if you want forward streamed response you should use different sinatra interface.

But I don't need streaming :thinking: Just request - response. Nothing special. That's why missed Content-Length header after upgrade makes me frustrated.

**Comment by dm1try** [comment]
> But I don't need streaming 

but the proxy client says that it has streaming capability(HTTP/1.1), and any component in the stack can decide to choose any delivering mechanism for the payload.  ex, `puma` handling https://github.com/puma/puma/blob/7a6593760d667dff95953e15c2327892e2da673c/lib/puma/server.rb#L721-L722

**Comment by nbulaj** [comment]
Well, while I disagree with _so, you should be prepared for this_ now I see the whole picture :)

So before there was a Content_Length header and our proxy client see requests finished (we have full body), but now we don't have this one and it waits for streaming. Thanks @dm1try, in a meanwhile we can review the proxy client setup

**Comment by dblock** [comment]
@dm1try Sounds like we want a spec around Content-Length and document the behavior, plus note in UPGRADING? @nbulaj care to contribute?

**Comment by nbulaj** [comment]
I don't have a time for it right now and I don't actually understand from what to start - don't sure how #1956 could break the headers :( Maybe @ioquatix have better vision for Rack and the changes

**Comment by dm1try** [comment]
> @dm1try Sounds like we want a spec around Content-Length

I'm not sure if the framework should be responsible for this. Users always can include `Rack::ContentLength` middleware in their stack(btw, @nbulaj this is another recipe), explicitly.

> So before there was a Content_Length header and our proxy client see requests finished (we have full body), but now we don't have this one and it waits for streaming.

nope, now your proxy builds the wrong response mixing `Transfer-Encoding: chunked` and building the response as usual with `Content-Length`, it does not waits anything, the clients of the proxy wait for the chunked response, as your proxy says that the response is chunked, but the forwarded response is built as not-chunked one.

> don't sure how #1956 could break the headers

it does not break headers, it fixes the problem where `grape  < 1.3.0` does not work with `rack => 2.1.0`. But upgrading rack imply other changes that I already referenced. So if you downgrade `rack` to `2.0.8` you will see the old behavior;
`grape` uses `Rack::Response` under the hood
https://github.com/ruby-grape/grape/blob/002280415a46b1cabea565533141298781a48553/lib/grape/middleware/base.rb#L61
`Rack::Response` in `rack 2.0.8` sets `Content-Length`
`Rack::Response` in `rack 2.1.0` does not set `Content-Length` because the interface was changed in https://github.com/rack/rack/commit/8c62821f4a464858a6b6ca3c3966ec308d2bb53e
see https://github.com/ruby-grape/grape/issues/2051#issuecomment-624927490

there's never been a specification for setting `Content-Length` header, so it does not look like a regression.

**Comment by nbulaj** [comment]
> Users always can include Rack::ContentLength middleware in their stack (btw, @nbulaj this is another recipe

It solved the issue. But for me it looks strange when I need to add `use Rack::ContentLength` so that broken code now works. 

> nope, now your proxy builds the wrong response mixing Transfer-Encoding: chunked and building the response as usual with Content-Length, it does not waits anything, the clients of the proxy wait for the chunked response, as your proxy says that the response is chunked, but the forwarded response is built as not-chunked one.

Yep, I was incorrect. That's a right point - mixed and wrong headers.

I also checked the gemfile.lock and found that  rack (2.0.8) became rack (2.2.2)

**Comment by dm1try** [comment]
> But for me it looks strange when I need to add use Rack::ContentLength so that broken code now works.

yep, it is just a workaround for your forwarding proxy.
this will prevent any other component in the stack(HTTP/1.1!) to use a different mechanism for delivering the payload as you know that your forwarding proxy does not support chunked responses. The same can be achieved by disabling "chunked" responses in your application server, in `puma`, etc.

RFC 2616
 > All HTTP/1.1 applications MUST be able to receive and decode the
   "chunked" transfer-coding, and MUST ignore chunk-extension extensions
   they do not understand.

**Comment by ioquatix** [comment]
Setting the `content-length` header from within the application is generally an error.

When the response body is an array of strings, it can be easily computed by the server, and we've seen application generated `content-length` headers become wrong after going through middleware which adds content, compresses content or does some other manipulation without updating the `content-length` and the server just accepting it and the client borking.

The only somewhat legitimate place to set a content-length header is when your response is not an array of strings, and you DO know how long it is going to be - i.e. a file. However, in that case, the server can almost always do a better job, e.g. using `to_path`.

**Comment by nbulaj** [comment]
Well all this situation got me sad...

Now with all the changes in Rack and other gems I need to decide - whether to use old HTTP 1.0 (but I don't want, I wanna 1/1 without streaming) to communicate with the main API or try to inject Content-Length somehow (calculate on main API? or in proxy?) and got an issues on compression or middleware manipulations in the future :(

**Comment by dm1try** [comment]
> all the changes in Rack and other gems

All those changes just show up [the bug](https://github.com/ruby-grape/grape/issues/2030#issuecomment-624772241) in the implementation of your forwarding proxy. Fix it first. Other suggestions is just a workaround to get an "old working" behavior.

**Comment by nbulaj** [comment]
Let's close the issue as I think it's not related to Grape

**Comment by dblock** [comment]
Thanks everyone for your comments!
